File: frontend/src/types/listing.ts
export type ListingStatus = "active" | "pending" | "sold" | "off_market";
export type ScoreColor = "green" | "blue" | "yellow" | "red";

export interface Scores {
  price_score: number | null;
  location_score: number | null;
  build_score: number | null;
  overall_score: number | null;
  color: ScoreColor | null;
}

export interface NeighborParcel {
  parcel_number: string | null;
  owner: string | null;
  acres: number | null;
  boundary: [number, number][] | null;
}

export interface Parcel {
  parcel_number: string | null;
  owner: string | null;
  tax_value: number | null;
  zoning: string | null;
  road_frontage: boolean | null;
  utilities: string | null;
  elevation_ft: number | null;
  /** "nc_onemap" for real statewide parcel data, "estimated" for the fallback stub. */
  data_source: "nc_onemap" | "estimated";
  boundary_coordinates: [number, number][] | null;
  neighbor_parcels: NeighborParcel[] | null;
}

export interface Soil {
  soil_type: string | null;
  perk_possible: boolean | null;
  soil_rating: number | null;
}

export interface Flood {
  flood_zone: string | null;
}

export interface Buildability {
  well_required: boolean | null;
  septic_required: boolean | null;
  estimated_site_cost: number | null;
}

export interface Utilities {
  electric: boolean | null;
  internet: boolean | null;
  gas: boolean | null;
}

export interface Distances {
  costco: number | null;
  whole_foods: number | null;
  walmart: number | null;
  cvs: number | null;
  hospital: number | null;
  lowes: number | null;
  home_depot: number | null;
  i85: number | null;
}

export interface Listing {
  id: string;
  provider: string;
  address: string | null;
  county: string | null;
  city: string | null;
  zipcode: string | null;
  latitude: number;
  longitude: number;
  price: number;
  acres: number;
  price_per_acre: number;
  status: ListingStatus;
  url: string | null;
  last_updated: string;
  scores: Scores | null;
}

export interface ListingDetail extends Listing {
  parcel: Parcel | null;
  soil: Soil | null;
  flood: Flood | null;
  buildability: Buildability | null;
  utilities: Utilities | null;
  distances: Distances | null;
}

export interface ListingFilters {
  county?: string;
  status?: ListingStatus;
  min_price?: number;
  max_price?: number;
  min_acres?: number;
  max_acres?: number;
  min_score?: number;
  road_frontage?: boolean;
  limit?: number;
  offset?: number;
}

export interface CountyBreakdown {
  county: string;
  count: number;
  avg_price: number;
}

export interface Dashboard {
  properties_found: number;
  average_price: number;
  average_price_per_acre: number;
  top_homestead_score: number | null;
  best_deal_listing_id: string | null;
  new_today: number;
  county_breakdown: CountyBreakdown[];
}

export interface CostEstimateInput {
  land_price: number;
  acres: number;
  needs_well?: boolean;
  needs_septic?: boolean;
  driveway_length_feet?: number;
  home_sqft?: number;
  solar_system_kw?: number;
  include_solar?: boolean;
  down_payment_pct?: number;
  annual_interest_rate?: number;
  term_years?: number;
  assessed_tax_value?: number;
}

export interface CostEstimateResult {
  land_purchase: number;
  site_clearing: number;
  grading: number;
  driveway: number;
  well: number;
  septic: number;
  electrical: number;
  survey: number;
  engineering: number;
  permits: number;
  construction_cost: number;
  solar: number;
  total_project_cost: number;
  down_payment: number;
  loan_amount: number;
  monthly_mortgage: number;
  monthly_taxes: number;
  monthly_insurance: number;
  monthly_pmi: number;
  total_monthly_payment: number;
}

export interface MapFeature {
  type: "Feature";
  geometry: { type: "Point"; coordinates: [number, number] };
  properties: {
    id: string;
    address: string | null;
    county: string | null;
    price: number;
    acres: number;
    price_per_acre: number;
    status: ListingStatus;
    overall_score: number | null;
    score_color: ScoreColor | null;
    flood_zone: string | null;
    soil_rating: number | null;
    perk_possible: boolean | null;
    estimated_site_cost: number | null;
    elevation_ft: number | null;
    /** "nc_onemap" for real statewide parcel data, "estimated" for the fallback stub. */
    parcel_data_source: "nc_onemap" | "estimated" | null;
    parcel_boundary: [number, number][] | null;
    neighbor_parcels: NeighborParcel[] | null;
  };
}

export interface MapData {
  type: "FeatureCollection";
  features: MapFeature[];
}

File: backend/app/schemas/listing.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, computed_field

from app.models.listing import ListingStatus


class NeighborParcelOut(BaseModel):
    parcel_number: str | None = None
    owner: str | None = None
    acres: float | None = None
    boundary: list[list[float]] | None = None


class ParcelOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    parcel_number: str | None
    owner: str | None
    tax_value: float | None
    zoning: str | None
    road_frontage: bool | None
    utilities: str | None
    elevation_ft: float | None
    data_source: str
    neighbor_parcels: list[NeighborParcelOut] | None

    # Raw PostGIS geometry isn't JSON-serializable; excluded from output and
    # converted to plain [lat, lon] pairs by `boundary_coordinates` below.
    boundary: Any = Field(default=None, exclude=True)

    @computed_field
    @property
    def boundary_coordinates(self) -> list[list[float]] | None:
        from app.services.geo import polygon_geometry_to_coordinates

        return polygon_geometry_to_coordinates(self.boundary)


class SoilOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    soil_type: str | None
    perk_possible: bool | None
    soil_rating: float | None


class FloodOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    flood_zone: str | None


class BuildabilityOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    well_required: bool | None
    septic_required: bool | None
    estimated_site_cost: float | None


class UtilitiesOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    electric: bool | None
    internet: bool | None
    gas: bool | None


class DistancesOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    costco: float | None
    whole_foods: float | None
    walmart: float | None
    cvs: float | None
    hospital: float | None
    lowes: float | None
    home_depot: float | None
    i85: float | None


class ScoresOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    price_score: float | None
    location_score: float | None
    build_score: float | None
    overall_score: float | None

    @computed_field
    @property
    def color(self) -> str | None:
        from app.scoring.engine import score_color

        return score_color(self.overall_score) if self.overall_score is not None else None


class ListingOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    provider: str
    address: str | None
    county: str | None
    city: str | None
    zipcode: str | None
    latitude: float
    longitude: float
    price: float
    acres: float
    price_per_acre: float
    status: ListingStatus
    url: str | None
    last_updated: datetime
    scores: ScoresOut | None = None


class ListingDetailOut(ListingOut):
    parcel: ParcelOut | None = None
    soil: SoilOut | None = None
    flood: FloodOut | None = None
    buildability: BuildabilityOut | None = None
    utilities: UtilitiesOut | None = None
    distances: DistancesOut | None = None


class ListingFilterParams(BaseModel):
    county: str | None = None
    status: ListingStatus | None = None
    min_price: float | None = None
    max_price: float | None = None
    min_acres: float | None = None
    max_acres: float | None = None
    min_score: float | None = None
    no_hoa: bool | None = None
    road_frontage: bool | None = None
    limit: int = 50
    offset: int = 0


class CountyBreakdown(BaseModel):
    county: str
    count: int
    avg_price: float


class DashboardOut(BaseModel):
    properties_found: int
    average_price: float
    average_price_per_acre: float
    top_homestead_score: float | None
    best_deal_listing_id: uuid.UUID | None
    new_today: int
    county_breakdown: list[CountyBreakdown]
